### ISSUE TITLE
Editor freezes when using replace in files for > 200 files, 

### ISSUE THREAD

**Issue by lencioni** [microsoft/vscode-eslint#940]
I am using the following configuration:

```json
 "[javascript]": {
    "editor.formatOnSave": false
  },
  "[javascriptreact]": {
    "editor.formatOnSave": false
  },
  "[typescript]": {
    "editor.formatOnSave": false
  },
  "[typescriptreact]": {
    "editor.formatOnSave": false
  },

  "editor.codeActionsOnSave": {
    "source.fixAll.eslint": true
  },
```

I noticed recently when doing large replace in files operations that my editor totally freezes up. After waiting a few minutes, I get a message saying that the window has become unresponsive.

I tried using `"eslint.codeActionsOnSave.mode": "problems",` to see if it would be better, but it is not.

VSCode stats:

Version: 1.43.2
Commit: 0ba0ca52957102ca3527cf479571617f0de6ed50
Date: 2020-03-24T07:34:57.037Z
Electron: 7.1.11
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Darwin x64 19.4.0

vscode-eslint v 2.1.2

**Comment by dbaeumer** [comment]
The problem here is that VS Code opens all these file in the editor, which makes them travel to the ESLint server which validates them. This all happens async so it shouldn't freeze the editor.

@bpasero what is the reason that we are still opening all these files. With @alexdima undo work could we simply not open these files anymore?

**Comment by dbaeumer** [comment]
I actually don't see what the ESLint extension can do to avoid this. I have no means to distinguish these cases (e.g. editor open versus find / replace open)

**Comment by bpasero** [comment]
@dbaeumer if we detect that a model changes and thus becomes dirty, we wait for a short delay before opening the model as editor. The rationale is that we do not want dirty decorations in the workbench (e.g. over the explorer) without associated opened editor. A dirty model should always result in a dirty editor.

Now, that delay is little (50ms) I think, so it could result in editors opening that are then marked as saved because the save took >50ms.

**Comment by dbaeumer** [comment]
@bpasero why do we not save the file after making the changes and don't open the file at all in the editor. To my knowledge one of the reasons for opening it in the editor was to be able to undo the find / replace. With Alex's work I guess we should be able to provide undo (e.g. Something like `Undo find/replace`) which works even if all the files got saved.

**Comment by bpasero** [comment]
@dbaeumer yes, that is what "Search & Replace" does and it does not result in editors opening to my knowledge.

**Comment by lencioni** [comment]
I've also noticed that it is slow when using Replace in files in general, even with smaller groups. In some cases, it will save the file in a broken state--like maybe ESLint didn't run autofix enough times to get the file into a stable state before it was actually saved.

**Comment by dbaeumer** [comment]
@bpasero you are right that it doesn't show the editor. But it opens a model, makes the changes, and closes the model which generates these events on the extension host side. I always thought we do that to get undo. But I guess that is not correct.

To make this a more pleasant experience we either need to:

- not open these models in a way that the extension host receives an event
- have access to the UI model in the extension host to know that these models are not shown in an editor and therefore ESLint shouldn't validate them.  The corresponding VS Code item is https://github.com/microsoft/vscode/issues/15178

I will mark the item as upstream for now since it can't be addressed without work in VS Code.

**Comment by lencioni** [comment]
Could this be related? https://github.com/microsoft/vscode-eslint/issues/1033

**Comment by dbaeumer** [comment]
@lencioni no it is unrelated since the scenario described here causes unnecessary load on the extension.

**Comment by dbaeumer** [comment]
VS Code is working on a tab model.

**Comment by lencioni** [comment]
@dbaeumer Cool, thanks for the update! Do you have any links that explain what that means or how it relates to this issue?

**Comment by dbaeumer** [comment]
The tab model will allow to avoid to sync these documents to the eslint server.

**Comment by dbaeumer** [comment]
Converting to pull model diagnostics will fix this.

**Comment by dbaeumer** [comment]
The conversion to pull model diagnostics has happen.
